server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java

```diff
@@ -1117,11 +1121,23 @@ public void parseXContent(XContentParser parser, boolean checkTrailingTokens) th
                 if (FROM_FIELD.match(currentFieldName, parser.getDeprecationHandler())) {
                     from(parser.intValue());
                 } else if (SIZE_FIELD.match(currentFieldName, parser.getDeprecationHandler())) {
-                    size = parser.intValue();
+                    int parsedSize = parser.intValue();
+                    // we treat -1 as not-set, but deprecate it to be able to later remove this funny extra treatment
+                    if (parsedSize != -1) {
+                        size(parsedSize);
+                    } else {
+                        deprecationLogger.deprecate(
+                            DeprecationCategory.API,
+                            "search-api-size-1",
+                            "Using search size of -1 is deprecated and will be removed in future versions. "
+                            + "Instead, don't use the `size` parameter if you don't want to set it explicitely."
+                        );
+                    }
```

Comment:
it is not an option to use the same setter for both and have the validation in it so it is not duplicated?

Reply:
I considered this but decided against it because it would have meant making the already stricted setter on SearchSourceBuilder more lenient (allowing -1). For Rest and search body parameters the result would be the same, but not e.g. for client side usage of the SearchSourceBuilder. I think not going back one step there is more important than avoiding the temporary duplication. But I'm happy to change that if I cannot convince you.